### DISCUSSION TITLE
Conditions for fields in field arrays

### DISCUSSION THREAD

**Discussion by yldshv** [data-driven-forms/react-forms#1426] [Need help] [answered]
Hello everyone, today I encountered a problem for which there might already be a solution, but I might not be able to find it.

Snippet:
```json
{
   "name":"list",
   "label":"List",
   "component":"object-list", //our name for a field-array
   "fields":[
      {
         "name":"status",
         "label":"Status",
         "component":"select",
         "clearedValue":null,
         "dataType":"integer",
         "options":[
            {
               "label":"Status 1",
               "value":1
            },
            {
               "label":"Status 2",
               "value":2
            },
            {
               "label":"Status 3",
               "value":3
            }
         ],
      },
      {
         "name":"test",
         "label":"Test",
         "component":"radio",
         "dataType":"boolean",
         "options":[
            {
               "label":"No",
               "value":false
            },
            {
               "label":"Yes",
               "value":true
            }
         ],
         "condition":{
            "or":[
               {
                  "when":"list[0].status", //just to highlight this
                  "is":1
               },
               {
                  "when":"list[0].status", //just to highlight this
                  "is":3
               }
            ],
            "then":{
               "visible":true
            }
         }
      }
   ]
}
```

As you can see here, I have list[0].status in the "when" field as a field name. Now, logically, the condition only works for the first item in the array list.

Is there a way to pass this differently in the schema, for example, list[index].status or something similar? If not, wouldn't that be a useful addition?

Thanks in advance and best regards,
Daniel

**Comment by Hyperkid123** [comment]
Hi @yldshv. You can define the condition as a function to be able to dynamically resolve the `when` attribute.

https://www.data-driven-forms.org/schema/condition-schema#when

For your  example I believe the function would look something similar to this:
```js
          condition: {
            or: [
              {
                when: (field) => {
                  return `${field.name.split(".").shift()}.status`;
                }, //dynamic when condition
                is: 1,
              },
              {
                when: "list[0].status", //static
                is: 3,
              },
            ],
            then: {
              visible: true,
            },
          },
```

If you need a schema that is strictly defined via json, we have recently added a condition mapper: https://www.data-driven-forms.org/mappers/condition-mapper#heading-conditionmapper

Edit: Here is a [codesandbox](https://codesandbox.io/p/sandbox/youthful-williams-5gpzm6?file=%2Fsrc%2FApp.js%3A56%2C1-72%2C13&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clpmarfgs00072v6knymdemwq%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clpmarfgr00032v6kdu026iwt%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clpmarfgr00042v6kzj227398%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clpmarfgr00062v6kel3us441%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clpmarfgr00032v6kdu026iwt%2522%253A%257B%2522id%2522%253A%2522clpmarfgr00032v6kdu026iwt%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clpmb0f6400022v6kqxmwo1iy%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A56%252C%2522startColumn%2522%253A1%252C%2522endLineNumber%2522%253A72%252C%2522endColumn%2522%253A13%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clpmb0f6400022v6kqxmwo1iy%2522%257D%252C%2522clpmarfgr00062v6kel3us441%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clpmarfgr00052v6khh87yfsc%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%2522%257D%255D%252C%2522id%2522%253A%2522clpmarfgr00062v6kel3us441%2522%252C%2522activeTabId%2522%253A%2522clpmarfgr00052v6khh87yfsc%2522%257D%252C%2522clpmarfgr00042v6kzj227398%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clpmarfgr00042v6kzj227398%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) with your example. I just had to use the default field array component.

**Reply by yldshv** [reply]
Hi @Hyperkid123 ,
we actually store the schemas in a database, which is why I tried it with the conditionMapper. It seems to work in principle, but only during the initial load. After that, when there's a value change in the status field, the function in the condition mapper is not executed anymore. Do you have any tips? Do i have to execute it somehow manually?

Here's the code in the sandbox:
[CodeSandBox](https://codesandbox.io/p/sandbox/distracted-moon-zvgwcs?file=%2Fsrc%2FApp.js%3A98%2C1&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clpmarfgs00072v6knymdemwq%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clpmarfgr00032v6kdu026iwt%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clpmarfgr00042v6kzj227398%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clpmarfgr00062v6kel3us441%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clpmarfgr00032v6kdu026iwt%2522%253A%257B%2522id%2522%253A%2522clpmarfgr00032v6kdu026iwt%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clpmb0f6400022v6kqxmwo1iy%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A98%252C%2522startColumn%2522%253A1%252C%2522endLineNumber%2522%253A98%252C%2522endColumn%2522%253A1%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clpmb0f6400022v6kqxmwo1iy%2522%257D%252C%2522clpmarfgr00062v6kel3us441%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clpmarfgr00052v6khh87yfsc%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%2522%257D%255D%252C%2522id%2522%253A%2522clpmarfgr00062v6kel3us441%2522%252C%2522activeTabId%2522%253A%2522clpmarfgr00052v6khh87yfsc%2522%257D%252C%2522clpmarfgr00042v6kzj227398%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clpmarfgr00042v6kzj227398%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

Edit: 
I've been testing a bit more now and found that when I use the condition mapper for "is" it works as desired with every ValueChange from the field in "when" (static). However, when I have a static value in "is" and a function via the condition mapper in "when" it only executes the function once and then never again. I might be doing something wrong, though.

**Reply by Hyperkid123** [reply]
@yldshv Sorry I was busy with work stuff. I'll take a look, seems like the `when` callback is not being triggered properly then
. I'll let you know ASAP.

**Reply by Hyperkid123** [reply, marked as answer]
OK, you can fix the by adding a values subscription. Your condition depends on a value from another field but the form is by default set to not listen only to field changes.

Can you add `subscription={{ values: true }}` prop to the FormRendered?

**Reply by yldshv** [reply]
> OK, you can fix the by adding a values subscription. Your condition depends on a value from another field but the form is by default set to not listen only to field changes.
> 
> Can you add `subscription={{ values: true }}` prop to the FormRendered?

That helped. Now works as it should.

**Reply by Hyperkid123** [reply]
@yldshv OK. This was strange to me because I remember I had written some code to specifically get around the subscription issue as the subscription can have a negative impact on performance in very large forms. And in fact, we have missed the mapped condition case. I'll open a PR to ensure the condition works as expected by default without having to change the subscription.

**Reply by yldshv** [reply]
@Hyperkid123 while you working on it maybe here is the next thing :O mappedAttributes in an or condition:

```json
              "condition": {
                "or": [
                  {
                    "mappedAttributes": {
                      "when": ["objListResolver"]
                    },
                    "is": 1
                  },
                  {
                    "mappedAttributes": {
                      "when": ["objListResolver"]
                    },
                    "is": 7
                  }
                ],
                "then": {
                  "visible": true
                }
              }
```

does not work. **Error in console:** Missing conditionMapper entry for objListResolver!
Without the or condition it works :/

**Reply by Hyperkid123** [reply]
@yldshv I've converted the topic into an issue: #1428 

I'll take a look at the or variant as well.

**Comment by yldshv** [comment]
> @yldshv I've converted the topic into an issue: #1428
> 
> I'll take a look at the or variant as well.

Hi @Hyperkid123,
updated to 3.22.1 and everything works fine! Thank you!
